### DISCUSSION TITLE
[ZK Stack] - Multi-prover generates proof problems

### DISCUSSION THREAD

**Discussion by Marius-BC-Dev** [zkSync-Community-Hub/zksync-developers#317] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [X] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.73.0-nightly

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

1. Can multiple prover clients be used to generate proofs per circuit?For a circuit of type 0, when I am generating proofs with two prover clients, one prover exits abnormally after accepting the data and the other one works fine.
2. What machine configuration does prover need to generate proofs quickly? I currently take 40-50 minutes to generate proofs .
  Machine Configuration:
  CPU: 48C
  GPU: 3090
  GPU RAM: 20G

### Expected Behavior

Jobs for a class of circuits can be produced proof with multiple prover.

### Current Behavior

Jobs for a class of circuits can only be generated proof with one prover.

### Prior Research

- [ ] Yes, I searched for similar issues and did not find any matching ones.

**Comment by robik75** [comment, marked as answer]
Hello,
running multiple prover-instances in parallel on the same GPU is not supported.
When a prover instance starts, it allocates as much GPU RAM as it can so if you try to run a second one in parallel, it will fail because all the memory was already taken by the first one.

**Reply by Marius-BC-Dev** [reply]
I am using multiple GPU servers to run prover, different circuits are fine.

**Comment by loriyamazaki** [comment]
Yes, it is possible to use multiple prover clients to generate proofs per circuit. However, it's important to ensure that the prover clients are properly configured and synchronized to handle the workload efficiently. The abnormal exit of one prover client may indicate a configuration issue or a resource limitation.

To generate proofs quickly, consider optimizing the following aspects:

CPU and GPU Utilization: Ensure that both CPU and GPU resources are fully utilized during the proof generation process. You can monitor resource usage using system monitoring tools and adjust the configuration accordingly.

Memory Management: Optimize memory usage to prevent bottlenecks during proof generation. Ensure that there is sufficient memory available for each prover client to handle the workload effectively.

Parallel Processing: Explore options for parallelizing proof generation tasks across multiple prover clients to leverage the full computational power of your hardware setup.

Prover Configuration: Check the configuration settings of each prover client to ensure they are properly configured and aligned with the requirements of your workload.

By optimizing these aspects and possibly fine-tuning the configuration of your prover clients, you should be able to generate proofs more efficiently and effectively utilize multiple prover clients for the task.

**Reply by Marius-BC-Dev** [reply]
Thank you! Do you have any suggestions? Or relevant configuration documentation?

**Comment by jodychvs** [comment]
Yes, it's possible to use multiple prover clients to generate proofs for a circuit. However, the abnormal exit of one prover client while the other works fine suggests a potential issue with the setup or configuration of the malfunctioning prover client. Regarding the machine configuration for faster proof generation, your current setup seems quite robust with a powerful CPU and GPU. However, to optimize performance further, ensure that your prover software is properly configured and that all hardware components are utilized efficiently during the proof generation process. #ProverClients #ProofGeneration #MachineConfiguration

**Reply by Marius-BC-Dev** [reply]
Thank you! Do you have a suggested configuration? Or is there a relevant config file I can refer to?

**Comment by jillalbritton** [comment]
It is possible to use multiple prover clients to generate proofs per circuit. Ensure all prover clients are configured correctly and have sufficient computational resources to handle the workload efficiently.

**Reply by Marius-BC-Dev** [reply]
Give me an example? I can't get it to work on my own.

**Comment by syringjerry** [comment]
You can interact with smart contracts on zkSync using the Python SDK by deploying contracts, calling their functions, and querying their states. You may refer to the zkSync Python SDK documentation and explore examples provided there for guidance.

**Comment by bryanbrega** [comment]
The user is asking if multiple prover clients can be used to generate proofs per circuit, and if there are any machine configuration requirements to generate proofs quickly. They are experiencing an issue where one prover client exits abnormally while the other works fine when generating proofs for a circuit of type 0. The user has a machine configuration of 48C CPU, 3090 GPU, and 20G GPU RAM, but it currently takes 40-50 minutes to generate proofs. The user has not provided a repository link or additional details. They have searched for similar issues but did not find any matching ones.
